In [200]:
import pandas as pd

## Extração dos Dados
Para iniciar sua análise, você precisará importar os dados da API da Telecom X. Esses dados estão disponíveis no formato JSON e contêm informações essenciais sobre os clientes, incluindo dados demográficos, tipo de serviço contratado e status de evasão.
1.   Carregar os dados diretamente da API utilizando Python.
2.   Converter os dados para um DataFrame do Pandas para facilitar a manipulação.

In [201]:
url = 'https://raw.githubusercontent.com/ingridcristh/challenge2-data-science/refs/heads/main/TelecomX_Data.json'

In [202]:
df = pd.read_json(url) #ler em formato Json

In [203]:
df.head() #visualizar os dados

customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

## Dicionário de dados


* customerID: número de identificação único de cada cliente
* Churn: se o cliente deixou ou não a empresa
* gender: gênero (masculino e feminino)
* SeniorCitizen: informação sobre um cliente ter ou não idade igual ou maior que 65 anos
* Partner: se o cliente possui ou não um parceiro ou parceira
* Dependents: se o cliente possui ou não dependentes
* tenure: meses de contrato do cliente
* PhoneService: assinatura de serviço telefônico
* MultipleLines: assisnatura de mais de uma linha de telefone
* InternetService: assinatura de um provedor internet
* OnlineSecurity: assinatura adicional de segurança online
* OnlineBackup: assinatura adicional de backup online
* DeviceProtection: assinatura adicional de proteção no dispositivo
* TechSupport: assinatura adicional de suporte técnico, menos tempo de espera
* StreamingTV: assinatura de TV a cabo
* StreamingMovies: assinatura de streaming de filmes
* Contract: tipo de contrato
* PaperlessBilling: se o cliente prefere receber online a fatura
* PaymentMethod: forma de pagamento
* Charges.Monthly: total de todos os serviços do cliente por mês
* Charges.Total: total gasto pelo cliente

## TASK 1: Conhecendo o Dataset

Agora que você extraiu os dados, é fundamental entender a estrutura do dataset e o significado de suas colunas. Essa etapa ajudará a identificar quais variáveis são mais relevantes para a análise de evasão de clientes.

1. ✅ Explorar as colunas do dataset e verificar seus tipos de dados.
2. ✅ Consultar o dicionário para entender melhor os significados das variáveis.
3. ✅ Identificar as colunas mais relevantes para a análise de evasão.




In [204]:
df.shape #verificar o total de linhas e colunas do df original

(7267, 6)

In [205]:
df.columns #verificar o nome das colunas

Index(['customerID', 'Churn', 'customer', 'phone', 'internet', 'account'], dtype='object')

In [206]:
df.info() #verificar as informações do DF

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   customerID  7267 non-null   object
 1   Churn       7267 non-null   object
 2   customer    7267 non-null   object
 3   phone       7267 non-null   object
 4   internet    7267 non-null   object
 5   account     7267 non-null   object
dtypes: object(6)
memory usage: 340.8+ KB


### Campo para verificar os dados de colunas aninhadas e verificar informações relavantes nelas

In [207]:
data_customer = pd.json_normalize(df['customer'])
data_phone = pd.json_normalize(df['phone'])
data_internet = pd.json_normalize(df['internet'])
data_account = pd.json_normalize(df['account'])

In [208]:
data_customer.head()

gender  SeniorCitizen Partner Dependents  tenure
0  Female              0     Yes        Yes       9
1    Male              0      No         No       9
2    Male              0      No         No       4
3    Male              1     Yes         No      13
4  Female              1     Yes         No       3

In [209]:
data_phone.head()

PhoneService MultipleLines
0          Yes            No
1          Yes           Yes
2          Yes            No
3          Yes            No
4          Yes            No

In [210]:
data_account.head()

Contract PaperlessBilling     PaymentMethod  Charges.Monthly  \
0        One year              Yes      Mailed check             65.6   
1  Month-to-month               No      Mailed check             59.9   
2  Month-to-month              Yes  Electronic check             73.9   
3  Month-to-month              Yes  Electronic check             98.0   
4  Month-to-month              Yes      Mailed check             83.9   

  Charges.Total  
0         593.3  
1         542.4  
2        280.85  
3       1237.85  
4         267.4

In [211]:
data_internet.head()

InternetService OnlineSecurity OnlineBackup DeviceProtection TechSupport  \
0             DSL             No          Yes               No         Yes   
1             DSL             No           No               No          No   
2     Fiber optic             No           No              Yes          No   
3     Fiber optic             No          Yes              Yes          No   
4     Fiber optic             No           No               No         Yes   

  StreamingTV StreamingMovies  
0         Yes              No  
1          No             Yes  
2          No              No  
3         Yes             Yes  
4         Yes              No

### Criação de um DF normalizado com todas as colunas separadas do DF original
Remoção das colunas aninhadas e junção das colunas geradas acima para um DF normalizado

In [212]:
df_base = df.drop(columns=['customer', 'phone', 'internet', 'account']) #removendo as colunas aninhadas
df_normalizado = pd.concat([df_base, data_customer, data_phone, data_internet, data_account], axis=1) #juntar tudo num dataframe aninhado

In [213]:
df_normalizado.head() #Dataframe principal para ser usado

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female              0     Yes        Yes       9   
1  0003-MKNFE    No    Male              0      No         No       9   
2  0004-TLHLJ   Yes    Male              0      No         No       4   
3  0011-IGKFF   Yes    Male              1     Yes         No      13   
4  0013-EXCHZ   Yes  Female              1     Yes         No       3   

  PhoneService MultipleLines InternetService  ... OnlineBackup  \
0          Yes            No             DSL  ...          Yes   
1          Yes           Yes             DSL  ...           No   
2          Yes            No     Fiber optic  ...           No   
3          Yes            No     Fiber optic  ...          Yes   
4          Yes            No     Fiber optic  ...           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No         Yes         Yes              No        One year   
1               No          No          No             Yes  Month-to-month   
2              Yes          No          No              No  Month-to-month   
3              Yes          No         Yes             Yes  Month-to-month   
4               No         Yes         Yes              No  Month-to-month   

  PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0              Yes      Mailed check            65.6          593.3  
1               No      Mailed check            59.9          542.4  
2              Yes  Electronic check            73.9         280.85  
3              Yes  Electronic check            98.0        1237.85  
4              Yes      Mailed check            83.9          267.4  

[5 rows x 21 columns]

## TASK 2: Verificando Inconsistências nos Dados

Neste passo, verifique se há problemas nos dados que possam afetar a análise. Fique atento a valores ausentes, duplicados, erros de formatação e inconsistências nas categorias. Esse processo é essencial para garantir que os dados estejam prontos para as próximas etapas.

In [214]:
df_normalizado.isnull() #Verificar se existe valores nulos

customerID  Churn  gender  SeniorCitizen  Partner  Dependents  tenure  \
0          False  False   False          False    False       False   False   
1          False  False   False          False    False       False   False   
2          False  False   False          False    False       False   False   
3          False  False   False          False    False       False   False   
4          False  False   False          False    False       False   False   
...          ...    ...     ...            ...      ...         ...     ...   
7262       False  False   False          False    False       False   False   
7263       False  False   False          False    False       False   False   
7264       False  False   False          False    False       False   False   
7265       False  False   False          False    False       False   False   
7266       False  False   False          False    False       False   False   

      PhoneService  MultipleLines  InternetService  ...  OnlineBackup  \
0            False          False            False  ...         False   
1            False          False            False  ...         False   
2            False          False            False  ...         False   
3            False          False            False  ...         False   
4            False          False            False  ...         False   
...            ...            ...              ...  ...           ...   
7262         False          False            False  ...         False   
7263         False          False            False  ...         False   
7264         False          False            False  ...         False   
7265         False          False            False  ...         False   
7266         False          False            False  ...         False   

      DeviceProtection  TechSupport  StreamingTV  StreamingMovies  Contract  \
0                False        False        False            False     False   
1                False        False        False            False     False   
2                False        False        False            False     False   
3                False        False        False            False     False   
4                False        False        False            False     False   
...                ...          ...          ...              ...       ...   
7262             False        False        False            False     False   
7263             False        False        False            False     False   
7264             False        False        False            False     False   
7265             False        False        False            False     False   
7266             False        False        False            False     False   

      PaperlessBilling  PaymentMethod  Charges.Monthly  Charges.Total  
0                False          False            False          False  
1                False          False            False          False  
2                False          False            False          False  
3                False          False            False          False  
4                False          False            False          False  
...                ...            ...              ...            ...  
7262             False          False            False          False  
7263             False          False            False          False  
7264             False          False            False          False  
7265             False          False            False          False  
7266             False          False            False          False  

[7267 rows x 21 columns]

In [215]:
df_normalizado.isnull().values.any() #se existe algum valor null lu na

np.False_

### Foi observado que o DF não possui valores nulos e nem duplicados

In [216]:
df_normalizado.isnull().sum() #verificar se o DF possui dados null

customerID          0
Churn               0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
Charges.Monthly     0
Charges.Total       0
dtype: int64

In [217]:
df_normalizado.duplicated().sum() #verificar se existe valores duplicados

np.int64(0)

### Foi observado que o DF não possui IDs repetidos

In [218]:
df_normalizado['customerID'].nunique() #verificar se existe ID repetido

7267

### Verificação de dados em branco '' ou ' '

In [219]:
df_normalizado.apply(lambda col: (col.astype(str).str.strip() == '').sum()) #verifica dados ausentes em branco

customerID            0
Churn               224
gender                0
SeniorCitizen         0
Partner               0
Dependents            0
tenure                0
PhoneService          0
MultipleLines         0
InternetService       0
OnlineSecurity        0
OnlineBackup          0
DeviceProtection      0
TechSupport           0
StreamingTV           0
StreamingMovies       0
Contract              0
PaperlessBilling      0
PaymentMethod         0
Charges.Monthly       0
Charges.Total        11
dtype: int64

In [220]:
dados_customerID_vazios = df_normalizado[df_normalizado['Churn'].astype(str).str.strip() == '']

### Foi observado que 224 Clientes estão sem dados de Churn, ou seja, não sabemos se saiu da empresa ou não

In [221]:
dados_customerID_vazios

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
30    0047-ZHDTW        Female              0      No         No      11   
75    0120-YZLQA          Male              0      No         No      71   
96    0154-QYHJU          Male              0      No         No      29   
98    0162-RZGMZ        Female              1      No         No       5   
175   0274-VVQOQ          Male              1     Yes         No      65   
...          ...   ...     ...            ...     ...        ...     ...   
7158  9840-GSRFX        Female              0      No         No      14   
7180  9872-RZQQB        Female              0     Yes         No      49   
7211  9920-GNDMB          Male              0      No         No       9   
7239  9955-RVWSC        Female              0     Yes        Yes      67   
7247  9966-VYRTZ        Female              0     Yes        Yes      31   

     PhoneService     MultipleLines InternetService  ...         OnlineBackup  \
30            Yes               Yes     Fiber optic  ...                   No   
75            Yes                No              No  ...  No internet service   
96            Yes                No             DSL  ...                  Yes   
98            Yes                No             DSL  ...                  Yes   
175           Yes               Yes     Fiber optic  ...                  Yes   
...           ...               ...             ...  ...                  ...   
7158          Yes               Yes             DSL  ...                  Yes   
7180           No  No phone service             DSL  ...                   No   
7211          Yes               Yes     Fiber optic  ...                   No   
7239          Yes                No              No  ...  No internet service   
7247          Yes                No              No  ...  No internet service   

         DeviceProtection          TechSupport          StreamingTV  \
30                     No                   No                   No   
75    No internet service  No internet service  No internet service   
96                     No                  Yes                   No   
98                     No                  Yes                   No   
175                   Yes                   No                  Yes   
...                   ...                  ...                  ...   
7158                   No                   No                   No   
7180                   No                   No                  Yes   
7211                   No                   No                   No   
7239  No internet service  No internet service  No internet service   
7247  No internet service  No internet service  No internet service   

          StreamingMovies        Contract PaperlessBilling  \
30                     No  Month-to-month              Yes   
75    No internet service        Two year              Yes   
96                     No        One year              Yes   
98                     No  Month-to-month               No   
175                   Yes        One year              Yes   
...                   ...             ...              ...   
7158                   No        One year              Yes   
7180                   No  Month-to-month               No   
7211                   No  Month-to-month              Yes   
7239  No internet service        Two year              Yes   
7247  No internet service  Month-to-month              Yes   

                  PaymentMethod Charges.Monthly  Charges.Total  
30    Bank transfer (automatic)           79.00          929.3  
75      Credit card (automatic)           19.90         1355.1  
96             Electronic check           58.75         1696.2  
98      Credit card (automatic)           59.90         287.85  
175   Bank transfer (automatic)          103.15        6792.45  
...                         ...             ...            ...  
7158               Mailed check           54.25          773.2  
71

### Foi Decidido remover esses clientes que estão com os valores de Churn ausentes, pois pode influenciar na análise

In [222]:
df_normalizado = df_normalizado[df_normalizado['Churn'].astype(str).str.strip() != '']

### Agora não existe nenhum cliente com dados de Churn em brancos ou null

In [223]:
df_normalizado.apply(lambda col: (col.astype(str).str.strip() == '').sum()) #verifica os dados ausentes em branco

customerID           0
Churn                0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
Charges.Monthly      0
Charges.Total       11
dtype: int64

### Agora temos que verificar como tratar o Charges.Total que possuem valores em branco